File: server/assets/solutions/ohtu2019/part4/IntJoukkoSovellus/src/main/java/ohtu/intjoukkosovellus/IntJoukko.java

package ohtu.intjoukkosovellus;

public class IntJoukko {

    public final static int KAPASITEETTI = 5, 
                            OLETUSKASVATUS = 5;  

    private int kasvatuskoko;     
    private int[] luvut;     
    private int alkioidenLkm;   

    public IntJoukko() {
        this(KAPASITEETTI, OLETUSKASVATUS);
    }

    public IntJoukko(int kapasiteetti) {
        this(kapasiteetti, OLETUSKASVATUS);
    }
       
    public IntJoukko(int kapasiteetti, int kasvatuskoko) {
        if (kapasiteetti < 0 || kasvatuskoko < 0 ) {
            throw new IndexOutOfBoundsException("Invalid argument");
        }

        luvut = new int[kapasiteetti];
        alkioidenLkm = 0;
        this.kasvatuskoko = kasvatuskoko;

    }

    public boolean lisaa(int luku) {
        if ( kuuluu(luku)) {
            return false;
        }
        
        luvut[alkioidenLkm] = luku;
        alkioidenLkm++;
        
        if (alkioidenLkm % luvut.length == 0) {
            kasvataTaulukkoa();
        }
        
        return true;
    }

    private void kasvataTaulukkoa() {
        int [] uusiLuvut = new int[alkioidenLkm + kasvatuskoko];
        for (int i = 0; i < luvut.length; i++) {
            uusiLuvut[i] = luvut[i];
        }

        luvut = uusiLuvut;
    }

    public boolean kuuluu(int luku) {
        return luvunIndeksi(luku) != -1;
    }

    private int luvunIndeksi(int luku) {
        for (int i = 0; i < alkioidenLkm; i++) {
            if (luku == luvut[i]) {
                return i;
            }
        }
        return -1;
    }

    public boolean poista(int luku) {
        int kohta = luvunIndeksi(luku);
     
        if (kohta<0) {
            return false;
        }
    
        for (int j = kohta; j < alkioidenLkm - 1; j++) {
            luvut[j] = luvut[j + 1];
        }
        
        alkioidenLkm--;

        return true;
    }

    public int mahtavuus() {
        return alkioidenLkm;
    }


    @Override
    public String toString() {
        String tuotos = "";

        if (alkioidenLkm > 0) {
            for (int i = 0; i < alkioidenLkm - 1; i++) {
                tuotos += luvut[i];
                tuotos += ", ";
            }
            tuotos += luvut[alkioidenLkm - 1];
        }
        
        return "{" + tuotos + "}";
    }

    public int[] toIntArray() {
        int[] taulu = new int[alkioidenLkm];
        for (int i = 0; i < taulu.length; i++) {
            taulu[i] = luvut[i];
        }
        return taulu;
    }
   

    public static IntJoukko yhdiste(IntJoukko a, IntJoukko b) {
        IntJoukko tulos = new IntJoukko();
        
        for (int luku : a.toIntArray()) {
                tulos.lisaa(luku);
        }
        
        for (int luku : b.toIntArray()) {
                tulos.lisaa(luku);
        }
        
        return tulos;
    }

    public static IntJoukko leikkaus(IntJoukko a, IntJoukko b) {
        IntJoukko tulos = new IntJoukko();
        
        for (int luku : a.toIntArray()) {
            if (b.kuuluu(luku)) {
               tulos.lisaa(luku); 
            }
                
        }
        return tulos;    
    }
    
    public static IntJoukko erotus (IntJoukko a, IntJoukko b) {
        IntJoukko tulos = new IntJoukko();
        
        for (int luku : a.toIntArray()) {
            if (!b.kuuluu(luku)) {
               tulos.lisaa(luku); 
            }
                
        }
        return tulos; 
    }
        
}

File: server/assets/solutions/ohtu2023/part5/int-joukko/src/int_joukko.py
KAPASITEETTI = 5
OLETUSKASVATUS = 5


class IntJoukko:
    # tämä metodi on ainoa tapa luoda listoja
    def _luo_lista(self, koko):
        return [0] * koko

    def __init__(self, kapasiteetti=KAPASITEETTI, kasvatuskoko=OLETUSKASVATUS):
        if not isinstance(kapasiteetti, int) or kapasiteetti < 0:
            raise Exception("Väärä kapasiteetti")

        if not isinstance(kapasiteetti, int) or kapasiteetti < 0:
            raise Exception("Väärä kasvatuskoko") 
            
        self.kapasiteetti = kapasiteetti
        self.kasvatuskoko = kasvatuskoko

        self.joukon_luvut = self._luo_lista(self.kapasiteetti)

        self.alkioiden_lkm = 0

    def _paikka(self, luku):
        for i in range(0, self.alkioiden_lkm):
            if luku == self.joukon_luvut[i]:
                return i
        return -1

    def kuuluu(self, n):
        return self._paikka(n) != -1

    def lisaa(self, lisattava_luku):
        if self.kuuluu(lisattava_luku):
            return False
        
        # ei tilaa, luodaan uusi lista luvuille
        if self.alkioiden_lkm == self.kapasiteetti -1:
            taulukko_old = self.joukon_luvut
            self.kapasiteetti = self.alkioiden_lkm + self.kasvatuskoko
            self.joukon_luvut = self._luo_lista(self.kapasiteetti)
            self._kopioi_lista(taulukko_old, self.joukon_luvut)

        self.joukon_luvut[self.alkioiden_lkm] = lisattava_luku
        self.alkioiden_lkm = self.alkioiden_lkm + 1

        return True

    def poista(self, poistettava_luku):
        kohta = self._paikka(poistettava_luku)

        if kohta == -1:
            return False
        
        for j in range(kohta, self.alkioiden_lkm - 1):
            self.joukon_luvut[j] = self.joukon_luvut[j + 1]

        self.alkioiden_lkm = self.alkioiden_lkm - 1
        return True

    def _kopioi_lista(self, mista, minne):
        for i in range(0, self.alkioiden_lkm):
            minne[i] = mista[i]

    def mahtavuus(self):
        return self.alkioiden_lkm

    def to_int_list(self):
        taulu = self._luo_lista(self.alkioiden_lkm)
        self._kopioi_lista(self.joukon_luvut, taulu)    
        return taulu

    @staticmethod
    def yhdiste(a, b):
        tulos = IntJoukko()

        for luku in a.to_int_list():
            tulos.lisaa(luku)
        for luku in b.to_int_list():
            tulos.lisaa(luku)

        return tulos

    @staticmethod
    def leikkaus(a, b):
        tulos = IntJoukko()

        for luku in a.to_int_list():
            if b.kuuluu(luku):
                tulos.lisaa(luku)

        return tulos

    @staticmethod
    def erotus(a, b):
        tulos = IntJoukko()

        for luku in a.to_int_list():
            if not b.kuuluu(luku):
                tulos.lisaa(luku)
                
        return tulos

    def __str__(self):
        if self.alkioiden_lkm == 0:
            return "{}"

        tuotos = ""
        for i in range(0, self.alkioiden_lkm - 1):
            tuotos = tuotos + str(self.joukon_luvut[i])  + ", "
        tuotos = tuotos + str(self.joukon_luvut[self.alkioiden_lkm - 1])
        return "{"+ tuotos + "}"
